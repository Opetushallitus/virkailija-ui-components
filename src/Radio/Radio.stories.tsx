import * as React from 'react';
import { storiesOf } from '@storybook/react';
import Radio from './index';
import { boolean } from '@storybook/addon-knobs';

const stories = storiesOf('Components/Radio', module);

stories.add('Default', () => {
  const disabled = boolean('Disabled', false) as boolean;
  const error = boolean('Error', false) as boolean;
  const checked = boolean('Checked', false) as boolean;
  const fullWidth = boolean('Full width', true) as boolean;

  return (
    <Radio
      disabled={disabled}
      error={error}
      fullWidth={fullWidth}
      checked={checked}
    >
      Radio
    </Radio>
  );
});
