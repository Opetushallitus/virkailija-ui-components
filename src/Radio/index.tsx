import * as React from 'react';
import styled, { css } from 'styled-components';
import { hideVisually } from 'polished';

import { disabledStyle } from '../system';

const RadioContainer = styled.div`
  position: relative;
  flex: 0;
  line-height: 0;
  top: 0.15em;
`;

const Icon = styled.svg`
  background-color: white;
  border-radius: 50%;
  width: 0.4em;
  height: 0.4em;
`;

const HiddenRadio = styled.input.attrs({ type: 'radio' })`
  ${hideVisually}
`;

const StyledRadio = styled.div<{ checked: boolean; error: boolean }>`
  position: relative;
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid ${({ theme }) => theme.colors.inputBorder};
  box-shadow: 0 0 0 0 transparent;
  border-radius: 50%;
  transition: box-shadow 0.25s, border-color 0.25s, background-color 0.25s;
  display: flex;
  justify-content: center;
  align-items: center;

  &:hover,
  ${HiddenRadio}:focus + & {
    border-color: ${({ theme, error }) =>
      error ? theme.colors.danger.main : theme.colors.primary.main};
  }

  ${HiddenRadio}:focus + & {
    box-shadow: 0 0 0 3px ${({ theme }) => theme.colors.primary.focusOutline};
  }

  ${({ checked }) =>
    checked &&
    css`
      background-color: ${({ theme }) => theme.colors.primary.main};
      border-color: ${({ theme }) => theme.colors.primary.main};
    `}

  ${({ error }) =>
    error &&
    css`
      border-color: ${({ theme }) => theme.colors.danger.main};

      &:hover,
      ${HiddenRadio}:focus + & {
        border-color: ${({ theme }) => theme.colors.danger.main};
      }

      ${HiddenRadio}:focus + & {
        box-shadow: 0 0 0 3px ${({ theme }) => theme.colors.danger.focusOutline};
      }
    `}

  ${Icon} {
    visibility: ${(props) => (props.checked ? 'visible' : 'hidden')};
  }
`;

const Label = styled.label<{ disabled: boolean; error: boolean }>`
  cursor: pointer;
  font-family: ${({ theme }) => theme.fonts.main};
  font-size: 1rem;
  display: inline-flex;
  line-height: 1.5;
  color: ${({ theme }) => theme.colors.text.primary};
  align-items: flex-start;

  &:hover {
    & ${StyledRadio} {
      border-color: ${({ theme, error }) =>
        error ? theme.colors.danger.main : theme.colors.primary.main};
    }
  }

  ${({ error, theme }) =>
    error && {
      color: theme.colors.danger.main,
    }}

  ${disabledStyle}
`;

const LabelWrapper = styled.div`
  flex: 1;
  margin-left: 9px;
`;

type RadioBaseProps = {
  disabled?: boolean;
  checked?: boolean;
  children?: React.ReactNode;
  error?: boolean;
  className?: string;
  fullWidth?: boolean;
};

export type RadioProps = RadioBaseProps &
  Omit<React.ComponentProps<typeof HiddenRadio>, keyof RadioBaseProps>;

const Radio = React.forwardRef<HTMLInputElement, RadioProps>(
  (
    {
      className,
      checked = false,
      children,
      error = false,
      disabled = false,
      ...props
    }: RadioProps,
    ref,
  ) => (
    <Label disabled={disabled} error={error} className={className}>
      <RadioContainer>
        <HiddenRadio
          checked={checked}
          disabled={disabled}
          ref={ref}
          {...props}
        />
        <StyledRadio checked={checked} error={error}>
          <Icon />
        </StyledRadio>
      </RadioContainer>
      {children && <LabelWrapper>{children}</LabelWrapper>}
    </Label>
  ),
);

export default Radio;
