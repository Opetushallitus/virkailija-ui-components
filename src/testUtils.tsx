import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { createTheme } from './createTheme';

const theme = createTheme();

const Providers = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const dummyDiv = document.createElement('div');
dummyDiv.setAttribute('data-nothing', 'true');

// Return rootElement from render() for easier snapshot testing
const customRender = (ui: React.ReactElement, options: any = {}) => {
  const result = render(ui, { wrapper: Providers, ...options });
  return {
    ...result,
    rootElement: result.container.firstChild || dummyDiv,
  };
};

export * from '@testing-library/react';

export { customRender as render };
