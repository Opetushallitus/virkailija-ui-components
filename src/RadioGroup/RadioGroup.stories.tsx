import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import RadioGroup from './index';
import Radio from '../Radio';

const change = action('change') as any;

const stories = storiesOf('Components/RadioGroup', module);

stories.add('Default', () => {
  const disabled = boolean('Disabled', false) as boolean;
  const error = boolean('Error', false) as boolean;

  return (
    <RadioGroup value="1" onChange={change} error={error} disabled={disabled}>
      <Radio value="1">Radio 1</Radio>
      <Radio value="2">Radio 2</Radio>
      <Radio value="3">Radio 3</Radio>
    </RadioGroup>
  );
});

stories.add('With options prop', () => {
  return (
    <RadioGroup
      value="1"
      options={[
        { value: '1', label: 'Radio 1' },
        { value: '2', label: 'Radio 2' },
        { value: '3', label: 'Radio 3' },
      ]}
      onChange={change}
    />
  );
});
