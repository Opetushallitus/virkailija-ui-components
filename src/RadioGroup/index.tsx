import * as React from 'react';
import { isArray } from 'lodash';
import styled, { css } from 'styled-components';
import Radio from '../Radio';

const Container = styled.div<{ isLast: boolean }>`
  ${({ isLast }) =>
    !isLast &&
    css`
      margin-bottom: 4px;
    `}
`;

type RadioGroupOption = { value: string; label: React.ReactNode };

type RadioGroupChild = React.ReactElement<{
  checked?: boolean;
  value: string;
  onChange?: (arg: any) => void;
  disabled?: boolean;
  error?: boolean;
}>;

export type RadioGroupProps = {
  children?: RadioGroupChild[];
  value: string;
  onChange?: (arg: any) => void;
  disabled?: boolean;
  error?: boolean;
  options?: RadioGroupOption[];
  getIsDisabled?: (value: string) => boolean;
};

export const RadioGroup = ({
  value,
  onChange,
  disabled = false,
  options,
  error = false,
  children: childrenProp,
  getIsDisabled = () => false,
}: RadioGroupProps) => {
  let children: React.ReactNode = null;

  if (childrenProp) {
    const validChildren = React.Children.toArray(childrenProp).filter((c) =>
      React.isValidElement(c),
    ) as RadioGroupChild[];

    const childrenCount = React.Children.count(validChildren);

    children = validChildren.map((child, index) => {
      const checked = value !== undefined && child?.props?.value === value;
      const element = React.cloneElement(child, {
        checked,
        onChange,
        disabled:
          disabled ||
          getIsDisabled(child.props?.value) ||
          child.props?.disabled,
        error,
      });

      return (
        <Container
          key={child?.props?.value}
          isLast={index === childrenCount - 1}
        >
          {element}
        </Container>
      );
    });
  } else if (isArray(options)) {
    children = options.map(({ value: optionValue, label }, index) => (
      <Container isLast={index === options.length - 1} key={optionValue}>
        <Radio
          checked={value !== undefined && value === optionValue}
          onChange={onChange}
          value={optionValue}
          error={error}
          disabled={disabled || getIsDisabled(optionValue)}
        >
          {label}
        </Radio>
      </Container>
    ));
  }

  return <>{children}</>;
};

export default RadioGroup;
