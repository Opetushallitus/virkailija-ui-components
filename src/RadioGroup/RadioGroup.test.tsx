import React from 'react';
import { render } from '../testUtils';
import Radio from '../Radio';
import RadioGroup from './index';

it('Renders a RadioGroup with 3 radio buttons', () => {
  const { rootElement } = render(
    <RadioGroup value="b" onChange={() => {}}>
      <Radio value="a">Radio A</Radio>
      <Radio value="b">Radio B</Radio>
      <Radio value="c">Radio C</Radio>
    </RadioGroup>,
  );
  expect(rootElement).toMatchSnapshot();
});
