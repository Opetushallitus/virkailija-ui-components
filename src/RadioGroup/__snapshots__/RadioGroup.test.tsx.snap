// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a RadioGroup with 3 radio buttons 1`] = `
.c2 {
  position: relative;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  line-height: 0;
  top: 0.15em;
}

.c8 {
  background-color: white;
  border-radius: 50%;
  width: 0.4em;
  height: 0.4em;
}

.c4 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c6 {
  position: relative;
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid #ced4da;
  box-shadow: 0 0 0 0 transparent;
  border-radius: 50%;
  -webkit-transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6:hover,
.c3:focus + .c6 {
  border-color: #0a789c;
}

.c3:focus + .c5 {
  box-shadow: 0 0 0 3px rgba(10,120,156,0.2);
}

.c6 .c7 {
  visibility: hidden;
}

.c10:hover,
.c3:focus + .c10 {
  border-color: #0a789c;
}

.c3:focus + .c5 {
  box-shadow: 0 0 0 3px rgba(10,120,156,0.2);
}

.c10 .c7 {
  visibility: visible;
}

.c1 {
  cursor: pointer;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1.5;
  color: #666;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c1:hover .c5 {
  border-color: #0a789c;
}

.c9 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-left: 9px;
}

.c0 {
  margin-bottom: 4px;
}

<div
  class="c0"
>
  <label
    class="c1"
  >
    <div
      class="c2"
    >
      <input
        class="c3 c4"
        type="radio"
        value="a"
      />
      <div
        class="c5 c6"
      >
        <svg
          class="c7 c8"
        />
      </div>
    </div>
    <div
      class="c9"
    >
      Radio A
    </div>
  </label>
</div>
`;
